save-demo/src/main/kotlin/com/saveourtool/save/demo/service/KubernetesService.kt

```diff
@@ -72,72 +75,117 @@ class KubernetesService(
             status == null -> DemoStatus.ERROR
             status == HttpStatusCode.OK -> DemoStatus.RUNNING
             status.isSuccess() -> DemoStatus.STARTING
             else -> DemoStatus.STOPPED
         }
     }
 
+    private suspend fun configureDemoAgent(demo: Demo, version: String, retryNumber: Int = RETRY_TIMES): StringResponse {
+        val pod = getPodByDemo(demo)
+        logger.info("Configuring save-demo-agent ${demo.projectCoordinates()}")
+        val configuration = DemoAgentConfig(
+            configProperties.agentConfig.demoUrl,
+            demo.toDemoConfiguration(version),
+            demo.toRunConfiguration(),
+        )
+        return demoAgentRequestWrapper(listOf("setup"), pod, kc) { url ->
+            sendConfigurationRequestRetrying(url, configuration, retryNumber)
+        }
+    }
+
     /**
-     * @param demo demo entity
-     * @return [Deferred] of url of pod with demo
+     * Performs run request with [demoRunRequest] params to [demo]
+     *
+     * @param demo [Demo] entity
+     * @param demoRunRequest params of [DemoRunRequest]
+     * @return [HttpResponse] if response is fetched, null if some error occurred
      */
-    fun getUrl(demo: Demo): Deferred<String> = scope.async { getPodUrl(demo) }
-
-    private fun configureDemoAgent(demo: Demo, version: String) = scope.async {
-        val url = "${getPodUrl(demo)}/configure"
-        logger.info("Configuring save-demo-agent by url: $url")
-        val requestStatus = httpClient.post(url) {
-            contentType(ContentType.Application.Json)
-            accept(ContentType.Application.Json)
-            setBody(
-                DemoAgentConfig(
-                    configProperties.agentConfig.demoUrl,
-                    demo.toDemoConfiguration(version),
-                    demo.toRunConfiguration(),
-                )
-            )
-        }.status
+    suspend fun sendRunRequest(
+        demo: Demo,
+        demoRunRequest: DemoRunRequest,
+    ): HttpResponse? {
+        val pod = getPodByDemo(demo)
+        val response = try {
+            demoAgentRequestWrapper(listOf("run"), pod, kc) { url ->
+                logger.info("Sending POST request with url $url")
+                httpClient.post {
+                    url(url)
+                    contentType(ContentType.Application.Json)
+                    accept(ContentType.Application.Json)
+                    setBody(demoRunRequest)
+                }
+            }
+        } catch (connectException: ConnectException) {
+            null
+        }
+        return response
+    }
 
-        if (requestStatus.isSuccess()) {
-            logAndRespond(HttpStatusCode.OK, logger::info) {
-                "Job successfully started."
+    private suspend fun sendConfigurationRequestRetrying(
+        url: Url,
+        configuration: DemoAgentConfig,
+        retryNumber: Int,
+    ): StringResponse {
+        val requestStatusAndErrors = retry(retryNumber) { iteration ->
```

Comment:
?
```suggestion
        val (requestStatus, errors) = retry(retryNumber) { iteration ->
```

Reply:
thus we won't access `.first` and `.second`

Reply:
Yes, sure